pendulum

```diff
@@ -415,6 +415,166 @@ EOF
 	return 0
 }
 
+
+# DNS functions
+
+#=======================================
+# Code for check DNS availability
+#=======================================
+
+# multiple lock
+m_lock () {
+    local _lock_name="$1"
+    local _val_to_loc="$2"
+
+    _sem="m_lock_$1"
+    _val="$2"
+
+    RV=$(print "SET ${_sem} ${_val} NX" | redis runtime )
+
+   
+    [[ "$RV" = "OK" ]] && return 0;
+
+    return 1
+}
+
+
+m_unlock () {
+    local _sem="m_lock_$1"
+    local _val="$2"
+#    print "DECRBY $_sem $_val "
+    eval "NV=$(print "DECRBY $_sem $_val " | redis runtime)"
+    func " m_unlock : NV = $NV"
+    
+    [[ $NV = "0" ]] && {
+	_tmp=$(print "DEL $_sem " | redis runtime )
+	return 0 }
+
+
+    func " m_unlock : after DECRBY NV = $NV"
+    return 0
+}
+
+
+
+search_for_a_dns () {
+    func "$! spawned search_for_a_dns R=[$R]"
+    #
+    sed 1q $R/share/dnscrypt-resolvers.csv > $R/share/dnscrypt-resolvers-grepped.csv
+    grep $servertype $R/share/dnscrypt-resolvers.csv >> $R/share/dnscrypt-resolvers-grepped.csv
+    
+    # prints a random line from the resolvers csv file
+    local _rndresolv="$(perl -wne '$a = $_ if rand $. < 1 ; END { print $a }' $R/share/dnscrypt-resolvers-grepped.csv)"
+    
+    func " $$   # check if DNS is reachable"
+    eval "${execmap[dnscrypt-proxy]} -L $R/share/dnscrypt-resolvers.csv -R ${_rndresolv[(ws:,:)1]} -t 1 -m 1"
+    
+    [[ $? = 0 ]] || {
+	act "$$ : no good news from ${_rndresolv}"
+	m_unlock check_and_select_dns 1
+	return 0
+    }
+    dnscrypt_name="${_rndresolv[(ws:,:)1]}"
+
+    func "dnscrypt_name = ${dnscrypt_name}"
+    # Saving on redis the dnscrypt_name
+    act "$! > GOOD NEWS from ${_rndresolv}"
+    print "SET dns-server-selected ${dnscrypt_name}" | redis dynamic
+    
+    rm -f $R/share/dnscrypt-resolvers-grepped.csv
+
+#    PID_LIST=$(print "GET dnscrypt-pid-child-list " | redis dynamic | tr '_' ' ')
+
+    PID_LIST=$(print "GET dnscrypt-pid-child-list " | redis dynamic)
+    pid_array=("${(@s/_/)PID_LIST}")
+    LOCK=$(print "GET lock_check_and_select_dns" | redis dynamic )
+
+    act " restarting dnscrypt-proxy "
+    stop dnscrypt-proxy
+    start dnscrypt-proxy
+    
+    func " unlocking check_and_select_dns [$LOCK]"
+    # in this moment we can unlock the semaphore if the pendulum spawn another team it don't change our  PID_LIST variable
+    m_unlock check_and_select_dns 1
+
+#    for i in $pid_array ; do
+#	echo $i
+#	ps aux|grep $i
+#	kill -9 ${i}
+#    done
+
+}
+
+
+ __is_dns_reachable() {
+     [[ $1 = "(nil)" ]] && { return 1 }
+     [[ $1 = "" ]] && { return 1 }
+     [[ $1 = " " ]] && { return 1 }
+
+     act " Checking if dns  [$1] is reachable"
+     eval "${execmap[dnscrypt-proxy]} -L $R/share/dnscrypt-resolvers.csv -R $1 -t 1 -m 1"
```

Comment:
also here. appearing three times so far makes me think it should be in a separate function.